plugins/inputs/socketstat/socketstat.go

```diff
@@ -0,0 +1,219 @@
+
+// Description returns a short description of the plugin
+func (ss *Socketstat) Description() string {
+	return "Gather indicators from established connections, using iproute2's `ss` command."
+}
+
+// SampleConfig returns sample configuration options
+func (ss *Socketstat) SampleConfig() string {
+	return `
+  ## ss can display information about tcp, udp, raw, unix, packet, dccp and sctp sockets
+  ## Specify here the types you want to gather
+  # socket_proto = [ "tcp", "udp" ]
+  ## The default timeout of 1s for ss execution can be overridden here:
+  # timeout = "1s"
+`
+}
+
+// Gather gathers indicators from established connections
+func (ss *Socketstat) Gather(acc telegraf.Accumulator) error {
+	// best effort : we continue through the protocols even if an error is encountered,
+	// but we keep track of the last error.
+	for _, proto := range ss.SocketProto {
+		out, e := ss.Lister(ss.cmdName, proto, ss.Timeout)
+		if e != nil {
+			acc.AddError(e)
+			continue
+		}
+		e = ss.parseAndGather(out, proto, acc)
+		if e != nil {
+			acc.AddError(e)
+			continue
+		}
+	}
+	return nil
+}
+
+func socketList(cmdName string, proto string, timeout config.Duration) (*bytes.Buffer, error) {
+	// Run ss for the given protocol, return the output as bytes.Buffer
+	args := []string{"-in", "--" + proto}
+	cmd := exec.Command(cmdName, args...)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	err := internal.RunTimeout(cmd, time.Duration(timeout))
+	if err != nil {
+		return &out, fmt.Errorf("error running ss -in --%s: ", err)
+	}
+	return &out, nil
+}
+
+func (ss *Socketstat) parseAndGather(data *bytes.Buffer, proto string, acc telegraf.Accumulator) error {
+	scanner := bufio.NewScanner(data)
+	tags := map[string]string{}
+	fields := make(map[string]interface{})
+
+	// ss output can have blank lines, and/or socket basic info lines and more advanced
+	// statistics lines, in turns.
+	// In all non-empty lines, we can have metrics, so we need to group those relevant to
+	// the same connection.
+	// To achieve this, we're using the flushData variable which indicates if we should add
+	// a new measurement or postpone it to a later line.
+
+	// The first line is only headers
+	scanner.Scan()
+
+	flushData := false
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == "" {
+			continue
+		}
+		words := strings.Fields(line)
+
+		var err error
+		if !ss.BeginsWithBlank.MatchString(line) {
+			// A line with no starting whitespace means we're going to parse a new connection.
+			// Flush what we gathered about the previous one, if any.
+			if flushData {
+				acc.AddFields(ss.Measurement, fields, tags)
+			}
+
+			// Delegate the real parsing to getTagsAndState, which manages various
+			// formats depending on the protocol.
+			tags, fields = getTagsAndState(proto, words, ss.Log)
+
+			// This line containted metrics, so record that.
+			flushData = true
+		} else {
```

Comment:
Please add a `continue` in the `if` condition and unindent the `else` path.

Reply:
I believe that if we did that, we could miss the last measurement, by not running:
```
        if flushData {
                acc.AddFields(ss.measurement, fields, tags)
        }
```
at the end of this function.

Reply:
I don't think so, because this code is outside of the for-loop, isn't it? To me it seems that all code below if is in the else path... Did I miss something?

Reply:
My bad, you were right. I also reverted the test, as I found thenegative test (`if !<condition>`) less intuitive to read.